lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,208 @@
+require 'rexml/document'
+require 'securerandom'
+require 'open3'
+require 'base64'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @stderr, @ps_process = Open3.popen3(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        code = make_ps_code(powershell_code, output_ready_event_name, timeout_ms)
+
+        out, err = exec_read_result(code, output_ready_event)
+
+        # Powershell adds in newline characters as it tries to wrap output around the display (by default 80 chars).
+        # This behavior is expected and cannot be changed, however it corrupts the XML e.g. newlines in the middle of
+        # element names; So instead, part of the XML is Base64 encoded prior to being put on STDOUT and in ruby all
+        # newline characters are stripped. Then where required decoded from Base64 back into text
+        out = REXML::Document.new(out.gsub(/\n/,""))
+
+        hash = {}
+        REXML::XPath.each(out, "//Property").map { |x| [x.attributes["Name"], x.text] }.each do | prop |
+          hash[prop[0]] = prop[1]
+        end
+
+        {
+          :exitcode     => hash['exitcode'].to_i,
+          :errormessage => hash['errormessage'].nil? ? nil : Base64.decode64(hash['errormessage']),
```

Comment:
`errormessage` and `stdout` are no longer ever nil ... but they can be empty strings.